src/corehost/cli/fx_reference.h

```diff
@@ -17,11 +17,11 @@ class fx_reference_t
         : fx_name(_X(""))
         , fx_version(_X(""))
         , fx_version_number()
-        , has_patch_roll_fwd(false)
-        , patch_roll_fwd(false)
-        , has_roll_fwd_on_no_candidate_fx(false)
+        , has_apply_patches(false)
+        , apply_patches(false)
+        , has_roll_forward(false)
         , use_exact_version(false)
-        , roll_fwd_on_no_candidate_fx((roll_fwd_on_no_candidate_fx_option)0)
+        , roll_forward(roll_forward_option::Disable)
```

Comment:
Isn't `use_exact_version` the same as disable? 
Remove `use_exact_version`, `has_apply_patches`, `apply_patches`, `has_roll_forward`
Add `Patch`/`ClosestPatch` to roll_forward_option
Set roll_forward default to `LatestMajor`

Reply:
I'd agree that the separate `use_exact_version` shouldn't be needed.

I think combining `apply_patches` might have interesting implications around merging / choosing the most constrained? e.g. say the additional values are
`Minor, applyPatches=false` = `MinorClosestPatch` (more restrictive than `Minor`, less restrictive than `LatestPatch`)
`Major, applyPatches=false` = `MajorClosestPatch` (more restrictive than `Major`, less restrictive than `LatestMinor`)

If merging the settings:
`Minor, applyPatches=true` (i.e. `Minor`)
`Major, applyPatches=false` (i.e. `MajorClosestPatch`)

With keeping the settings separate, it would be `Minor, applyPatches=false`.
With the settings kept as one, it would be `Minor`

(Or I could be misunderstanding the suggestion and/or how things get merged)

Reply:
@elinor-fung My understanding was the `MinorClosestPatch` and `MajorClosestPatch` didn't exist.  Apply patch was ignored for these cases...

Reply:
Oh, I thought `applyPatches` was more of a modifier applicable to `LatestPatch`, `Minor`, and `Major`.

Reply:
`applyPatches` is a modifier - at least that's how it worked in 2.* and it exists in the code solely for backward compatibility reasons. I tried hard to not change the behavior for 2.* in any way. I was only partially successful:
* For release versions, the behavior is almost identical as before (one special case where it used to fail and now it works, and then the better retry logic)
* For pre-release versions the changes are much bigger. That said pre-release backward compat is sort of moot
  * Apps which use 2.* pre-release versions should be very rare by now
  * New apps which will use 3.0 pre-release will get the new behavior and should be perfectly fine with it.
  * Potential new apps which target 3.0 pre-release and use old-style `rollForwardOnNoCandidateFx` or `applyPatches` should be even rarer and the behavior there is different but still reasonable.

To keep the 2.* release to work as close to previous as possible, the code needs to keep `applyPatches` separate from `rollForward` exactly because the `applyPatches` acts as a modifier. It basically says: Once you're done picking version using the `rollForward` rules, normally you should then roll forward to latest patch of that version, but `applyPatches` may disable that behavior.
In theory it could be incorporated into the `rollForward` setting the way @elinor-fung  suggested, but also as she suggested, it introduces rather weird unification rules. Please note, that the unification rules are from the 2.* implementation so we don't have freedom in changing them (well unless we would be OK with breaking changes).

Reply:
> It basically says: Once you're done picking version using the rollForward rules, normally you should then roll forward to latest patch of that version, but applyPatches may disable that behavior.

I think we should remove this functionality. It makes no sense from a servicing perspective.
Especially since `applyPatch = false` is exceptionally rare.

We could map it `Disable` which also makes no sense.

Reply:
The problem is that if we do so, we effectively break that feature for everybody - even old apps - completely. I know it is against servicing, but that's exactly what the feature was for.
Same goes for `Disable` - it is a special case to be used in rare occasions where you really want to stick to a very specific version.
I think we need to support these "Escape hatches", but not recommend them.

The `applyPatches` is one we've shipped a while ago... and while I may not agree with how it works, I'm hesitant to break it completely.

Reply:
Perhaps `Disable` and `applyPatch=false` should both map to `ClosestPatch` which means we can rollforward `Patch` to avoid failing.

I thought we were talking about removing older patches from machines when we installed new ones anyway.

Reply:
For consistency `Patch` should definitely be renamed `LatestPatch`

Reply:
Removing older patches when new one is installed is planned for 3.0 if I remember correctly. So it doesn't apply to 2.*.

Where do you see just `Patch` - the enum uses `LatestPatch`.

Reply:
> It basically says: Once you're done picking version using the rollForward rules, normally you should then roll forward to latest patch of that version, but applyPatches may disable that behavior.

This makes no sense.  In no way could the App predict that it would allow rolling forward `major` or `minor`, but applying patches would be acceptable. We certainly shouldn't treat it that way in 3.*.  If we do it is a major new bug we are introducing.

Reply:
This part of the spec is completely wrong

+ `Minor` - if the exact `major.minor` is found, then the equal or lowest higher patch version is used. Otherwise the lowest higher minor is found and the lowest patch is used.
+ `Major` - if the exact `major` is found, the rules for `Minor` above are followed. Otherwise the lowest higher `major.minor.patch` is selected (so lowest patch available for a given `major.minor`).

Is should be
+ `Minor` - `applyPatches=false`, Behavior `Minor` behavior is unchanged.
+ `Major` - `applyPatches=false`, Behavior `Major` behavior is unchanged.

Reply:
Moved discussion to #5813

Reply:
> Removing older patches when new one is installed is planned for 3.0 if I remember correctly. So it doesn't apply to 2.*.

Maybe we should consider servicing 2.*.

> Where do you see just Patch - the enum uses LatestPatch.

I guess was remembering incorrectly.